### PR TITLE
cURL: Support all --data flags and fix parsing of key=value pairs

### LINKED ISSUE #5400: Insomnia does not parse curl request correctly
### Expected Behavior

The form name and value should be seperated:

image

### Actual Behavior

Both the form name and value are in the same place:

image

### Reproduction Steps

- Create new request
- Paste below curl command into request url box:

  ```text
  curl --request POST https://login.microsoftonline.com/domain.onmicrosoft.com/oauth2/v2.0/token --header "Content-Type: application/x-www-form-urlencoded" --data-urlencode "client_id=9154f684-a955-44a3-b577-8f9134502248" --data-urlencode "scope=https://graph.microsoft.com/.default" --data-urlencode "client_secret=804237f3-1c05-44aa-8bd3-4e38a90fdbf6" --data-urlencode "grant_type=client_credentials"
   ```

### Is there an existing issue for this?

- [X] I have searched the [issue tracker](https://www.github.com/Kong/insomnia/issues) for this problem.

### Additional Information

_No response_

### Insomnia Version

2022.7.0-beta.4

### What operating system are you using?

Windows

### Operating System Version

Windows 10 Pro Version 21H2

### Installation method

Download from github release

### Last Known Working Insomnia version

_No response_

### LINKED ISSUE #5696: Small Bug when pasting cURL into the URL field
When pasting a cURL command with URL encoded form data into the insomnia URL field, then the form parameters are automatically added. But key=values are not correctly parsed:

### How to reproduce

Past the following cURL into the URL field in insomnia client

````
curl --location --request POST 'https://domain.com/exampleDummy' \
  --header 'Accept: application/json' \
  --header 'Content-Type: application/x-www-form-urlencoded' \
  --data-urlencode 'response_type=code' \
  --data-urlencode 'state=1234abcd' 
````

### Expected behaviour
The URL is parsed and the --data-urlencode parameters are parsed as key and value and inserted into the correct columns in insomnia Form tab.

### Actual behaviour

Both "key=value" is parsed into the "key" column of the Form tab. The "value" column stays empty. See screenshot:

image

### Versions
Insomnia Native Client under Windows 2022.7.3
Build date: 16.1.2023
OS: Windows_NT x64 10.0.19044
Electorn: 22.0.0
Node: 16.17.1
V8: 10.8.168.20-electron.0
Architecture: x64

### PR DESCRIPTION
**Pull request by NickHackman**
changelog(Improvements): Improved parsing for cURL imports, supporting different data types.

Closes #5696
Closes #5400 

## Changes

- Supports `key=value` in `--data-urlencode`
- Supports all `data` types and handling their specific cases
- Supports files and their patterns
- Improved testing suite to make it clear what is supported and enabling easy extension in the future
- Documented added code with related documentation links

Implementation follows https://curl.se/docs/manpage.html

### CODE REVIEW

File: packages/insomnia/src/utils/importers/importers/curl.test.ts [hunk 1]
```diff
@@ -2,57 +2,155 @@ import { describe, expect, it } from '@jest/globals';
...
+import { Parameter } from '../entities';
 
 describe('curl', () => {
-  const url = 'http://example.com';
-  const method = '-X POST';
-  const mimeType = 'application/x-www-form-urlencoded';
-  const header = `-H 'Content-Type: ${mimeType}'`;
-
-  describe('encoding', () => {
+  describe('cURL --data flags', () => {
```
Comment (NickHackman) [inline packages/insomnia/src/utils/importers/importers/curl.test.ts:8]: Revamped the tests adding explicit checking for different command flags and arguments.

`-d`, `--data`, `--data-raw`, `--data-ascii`, `--data-binary`, and `--data-urlencode`

File: packages/insomnia/src/utils/importers/importers/curl.ts [hunk 2]
```diff
@@ -260,6 +234,110 @@ const importCommand = (parseEntries: ParseEntry[]): ImportRequest => {
...
+  /**
+   * https://curl.se/docs/manpage.html#--data-binary
+   */
+  'data-binary',
+
+  /**
+   * https://curl.se/docs/manpage.html#--data-ascii
+   */
+  'data-ascii',
+];
```
Comment (NickHackman) [inline packages/insomnia/src/utils/importers/importers/curl.ts:266]: Document with links to their sections in the man page

File: packages/insomnia/src/utils/importers/importers/curl.ts [hunk 3]
```diff
@@ -260,6 +234,110 @@ const importCommand = (parseEntries: ParseEntry[]): ImportRequest => {
...
+
+/**
+ * Parses pairs supporting only flags dictated by {@link dataFlags}
+ *
+ * @param keyedPairs pairs with cURL flags as keys.
+ */
+const pairsToDataParameters = (keyedPairs: PairsByName): Parameter[] => {
```
Comment (NickHackman) [inline packages/insomnia/src/utils/importers/importers/curl.ts:273]: Extracted out `textBodyParams` section, `Parameter[]` are just generated in this method.

File: packages/insomnia/src/utils/importers/importers/curl.ts [hunk 4]
```diff
@@ -260,6 +234,110 @@ const importCommand = (parseEntries: ParseEntry[]): ImportRequest => {
...
+const pairToParameters = (pair: Pair, allowFiles = false): Parameter[] => {
+  if (typeof pair === 'boolean') {
+    return [{ name: '', value: pair.toString() }];
+  }
+
+  return pair.split('&').map(pair => {
+    if (pair.includes('@') && allowFiles) {
+      const [name, fileName] = pair.split('@');
+      return { name, fileName, type: 'file' };
+    }
```
Comment (NickHackman) [inline packages/insomnia/src/utils/importers/importers/curl.ts:330]: Following the man pages from cURL, it's only ever documented in `--data-urlencode` that you can specify files via

`name@filename` syntax, this has the consequence of all `--data` flags supporting that format for code simplicity.

File: packages/insomnia/src/utils/importers/importers/curl.ts [hunk 5]
```diff
@@ -260,6 +234,110 @@ const importCommand = (parseEntries: ParseEntry[]): ImportRequest => {
...
+      continue;
+    }
+
+    switch (flagName) {
+      case 'd':
+      case 'data':
+      case 'data-ascii':
+      case 'data-binary':
+        dataParameters = dataParameters.concat(pairs.flatMap(pair => pairToParameters(pair, true)));
+        break;
```
Comment (NickHackman) [inline packages/insomnia/src/utils/importers/importers/curl.ts:289]: All of these behave the same in this context, as they all support files (technically in different ways)

File: packages/insomnia/src/utils/importers/importers/curl.ts [hunk 6]
```diff
@@ -260,6 +234,110 @@ const importCommand = (parseEntries: ParseEntry[]): ImportRequest => {
...
+ */
+const pairToParameters = (pair: Pair, allowFiles = false): Parameter[] => {
+  if (typeof pair === 'boolean') {
+    return [{ name: '', value: pair.toString() }];
+  }
+
+  return pair.split('&').map(pair => {
+    if (pair.includes('@') && allowFiles) {
+      const [name, fileName] = pair.split('@');
+      return { name, fileName, type: 'file' };
```
Comment (NickHackman) [inline packages/insomnia/src/utils/importers/importers/curl.ts:329]: Previously the `type` file was never specified for `--data` flags.

File: packages/insomnia/src/utils/importers/importers/curl.ts [hunk 7]
```diff
@@ -260,6 +234,110 @@ const importCommand = (parseEntries: ParseEntry[]): ImportRequest => {
...
+    switch (flagName) {
+      case 'd':
+      case 'data':
+      case 'data-ascii':
+      case 'data-binary':
+        dataParameters = dataParameters.concat(pairs.flatMap(pair => pairToParameters(pair, true)));
+        break;
+      case 'data-raw':
+        dataParameters = dataParameters.concat(pairs.flatMap(pair => pairToParameters(pair)));
+        break;
```
Comment (NickHackman) [inline packages/insomnia/src/utils/importers/importers/curl.ts:292]: `data-raw` is just `data`, but doesn't parse the `@` specially. So it just doesn't support files at all.

File: packages/insomnia/src/utils/importers/importers/curl.ts [hunk 8]
```diff
@@ -260,6 +234,110 @@ const importCommand = (parseEntries: ParseEntry[]): ImportRequest => {
...
+
+            return {
+              ...parameter,
+              value: encodeURIComponent(parameter.value ?? ''),
+            };
+          }));
+        break;
+      default:
+        throw new Error(`unhandled data flag ${flagName}`);
+    }
```
Comment (NickHackman) [inline packages/insomnia/src/utils/importers/importers/curl.ts:308]: This should never happen unless a new flag was added to `dataFlags`, but never handled here.

**Review by NickHackman — COMMENTED** [review]
See comments for implementation details and specific notes 😄

**Comment by NickHackman** [comment]
Hello @filfreire (tagging you because you've been active on tagged issues), this PR is open to review whenever is convenient for you and your team 🚀 

Please let me know if there's commits I need to rebase or required changes (also if I missed anything in the contributing guidelines). Happy to make any changes here, just trying to help resolve some open issues and give back to the project 😃

**Comment by filfreire** [comment]
Thanks for contributing @NickHackman, we'll look into this asap. cc @Kong/team-insomnia

**Comment by NickHackman** [comment]
"Test/OS (windows-latest)" failed due to a connection error

> npm ERR! network Invalid response body while trying to fetch https://registry.npmjs.org/check-engine: read ECONNRESET

**Review by jackkav — APPROVED** [review]
wow, awesome work thanks @NickHackman

**Review by filfreire — APPROVED** [review]

**Comment by filfreire** [comment]
@NickHackman you can claim a free tshirt for this PR contribution - find more at https://konghq.com/community/open-source-contribution

**Pull request merged**